pyTigerGraph/schema.py

```diff
@@ -0,0 +1,268 @@
+                                     ("is_directed", bool, e_type["IsDirected"]),
+                                     ("reverse_edge", str, e_type["Config"].get("REVERSE_EDGE"))],
+                                    bases=(Edge,), repr=False)
+                self._edge_types[e_type["Name"]] = e
+
+    def create_vertex_type(self, vertex: Vertex, outdegree_stats=True):
+        if vertex.__name__ in self._vertex_types.keys():
+            raise TigerGraphException(vertex.__name__+" already exists in the database")
+        if vertex.__name__ in self._vertex_edits.keys():
+            warnings.warn(vertex.__name__ + " already in staged edits. Overwriting previous edits.")
+        gsql_def = "ADD VERTEX "+vertex.__name__+"("
+        attrs = vertex.attributes
+        try:
+            primary_id = vertex.primary_id
+            primary_id_type = vertex.primary_id_type
+            primary_id_as_attr = vertex.primary_id_as_attribute
+        except:
+            raise TigerGraphException("Primary ID not defined. Run Vertex.define_primary_id() to define the Primary ID")
+        gsql_def += "PRIMARY_ID "+primary_id+" "+_py_to_tg_type(primary_id_type)
+        for attr in attrs.keys():
+            if attr == primary_id:
+                continue
+            else:
+                gsql_def += ", "
+                gsql_def += attr + " "+_py_to_tg_type(attrs[attr])
+        gsql_def += ")"
+        if outdegree_stats:
+            gsql_def += " WITH STATS='OUTDEGREE_BY_EDGETYPE'"
+        if outdegree_stats and primary_id_as_attr:
+            gsql_def += ", "
+        if primary_id_as_attr:
+            gsql_def += "PRIMARY_ID_AS_ATTRIBUTE='true'"
+        gsql_def += ";"
+        self._vertex_edits["ADD"][vertex.__name__] = gsql_def
+
+    def create_edge_type(self, edge: Edge, directed: bool=True, reverse_edge: Union[str, bool]=True):
+        if edge in self._edge_types.values():
+            raise TigerGraphException(edge.__name__+" already exists in the database")
+        if edge in self._edge_edits.values():
+            warnings.warn(edge.__name__ + " already in staged edits. Overwriting previous edits")
+        gsql_def = ""
+        if directed:
+            gsql_def += "ADD DIRECTED EDGE "+edge.__name__+"("
+        else:
+            gsql_def += "ADD UNDIRECTED EDGE "+edge.__name__+"("
+
+        # TODO: Handle Unions for multiple vertex types given an edge type
+        from_vert = edge.attributes["from_vertex"].__name__
+        to_vert = edge.attributes["to_vertex"].__name__
+        gsql_def += "FROM "+from_vert+", "+"TO "+to_vert
+        attrs = edge.attributes
+        for attr in attrs.keys():
+            if attr == "from_vertex" or attr == "to_vertex" or attr == "is_directed" or attr == "reverse_edge":
+                continue
+            else:
+                gsql_def += ", "
+                gsql_def += attr + " "+_py_to_tg_type(attrs[attr])
+        gsql_def += ")"
+
+        print(gsql_def)
```

Comment:
Is this intended for debugging or prod? Given the same comment above, I am wondering do we want a "show gsql" function or something like that?

Reply:
This is debugging, I was going to edit this in future stories, as the creation/deletion of types is in the backlog. Got ahead of myself here...